src/storage-client/src/storage_collections.rs

```diff
@@ -1554,6 +1554,7 @@ where
                         // as far the dependent's upper.
                         mz_ore::soft_assert_or_log!(
                             write_frontier.elements() == &[T::minimum()]
+                                || write_frontier.is_empty()
                                 || PartialOrder::less_than(&dependency_since, write_frontier),
```

Comment:
I'm not fully understanding the context of this check, but reading the comment above I wonder why this isn't using `PartialOrder::less_equal`? The comment says that the dependency since can not be "in advance of"/"past" the upper, both of which I think mean that the since cannot be _greater_ than the upper.

Is them being equal fine? If so, we can change the check to `less_equal` and can drop the `write_frontier.is_empty()` check. If it's not fine, can we update the comment to say why?

Reply:
It's using "advanced" they way timely/dd often usex "beyond", where

> "A time is beyond a frontier" == frontier.less_equal(time)

Though I think that meaning is less clearly defined/used when talking about two frontiers, maybe.

And it's important here that the since of dependencies _not_ be beyond the upper of the dependent. In practice, the dependency here will be a remap shard and the dependent will be a source/source exports. If the since is beyond (including if it's equal) we cannot differentiate between updates that have already been written to the source/export shard and new ones.

We try and uphold this requirement by using a `step_back` read policy here, for example: https://github.com/MaterializeInc/materialize/blob/f55491fa36fa5a9d18cd6dc4f7312a38c677b5c3/src/storage-controller/src/lib.rs#L763

Reply:
> It's using "advanced" they way timely/dd often usex "beyond", where

I was aware of that usage of "beyond" but didn't know that "in advance of" and "past" mean the same thing. Is there any word for `frontier.less_than(time)`? "Strictly in advance of"?

> If the since is beyond (including if it's equal) we cannot differentiate between updates that have already been written to the source/export shard and new ones.

Makes sense! Would be cool if you could update the comment to also say this. The current explanation ("If the  dependency since advanced past the dependent's upper, the dependent cannot read data from the dependency at its upper.") would not require that the since is strictly less than the upper.

Also could you update the comment to explain the new `write_frontier.is_empty()` edge case?

Reply:
pushed the change! ptal 🙏